File: schema-validation/src/validate_config.py
from asdf.schema import ValidationError, validate
import os
from rich.color import Color
from rich.console import Console
from rich.highlighter import Highlighter
from rich.style import Style
import sys
import textwrap
import yaml

# Highlights all text in red
class ErrorHighlighter(Highlighter):
    def highlight(self, text):
        text.stylize(Style(color=Color.from_rgb(255,0,0)), 0, len(text))

class SuccessHighlighter(Highlighter):
    def highlight(self, text):
        text.stylize(Style(color=Color.from_rgb(0,255,0)), 0, len(text))

# Force terminal so that colours are printed in the github action runner
errorConsole = Console(highlighter=ErrorHighlighter(), force_terminal=True)
console = Console(highlighter=SuccessHighlighter(), force_terminal=True)

def validate_config(environment: str, folder: str, directory_path: str, shared_buckets_path: str) -> None:
    """
    Validate if the config.yaml under the @directory_path is valid.
    Only uses @environment and @folder arguments to print more informative logs.

    Parameters:
        @environment: The dapla team environment the source belongs to i.e. 'uh-varer-prod' or 'play-obr-test'
        @folder: The statistics product folder which contains the configuration file i.e. 'ledstill' or 'sykefra'
        @directory_path: The filepath to the config.yaml file containing 'delomaten' configuration
        @shared_buckets_path: The filepath to the shared-buckets iam.yaml file containing the teams' shared-buckets

    """
    if not os.path.exists(directory_path):
        errorConsole.print(f'\n\nThe given directory path "{directory_path}" does not exist')
        sys.exit(1)

    config_data_path: str = os.path.join(directory_path, 'config.yaml')

    # We expect to find a config.yaml file in the directory
    if not os.path.exists(config_data_path):
        errorConsole.print(
            f'\n\nNo "config.yaml" file exists in the product source folder "{os.path.join(environment, folder)}"\n\n'
        )
        sys.exit(1)

    if not os.path.exists(shared_buckets_path):
        errorConsole.print(f'\n\nThe given shared-buckets path "{shared_buckets_path}" does not exist')
        sys.exit(1)

    with open(config_data_path) as stream:
        config_data: object = yaml.safe_load(stream.read())

    with open('./config_schema.yaml') as stream:
        config_schema: object = yaml.safe_load(stream.read())

    contextual_path = os.path.join(environment, folder, 'config.yaml')
    try:
        validate(config_data, schema=config_schema)
    except ValidationError as e:
        errorConsole.print(f'\n\nThe delomaten configuration file "{contextual_path}" is invalid:\n\n{e}\n\n')
        sys.exit(1)

    with open(shared_buckets_path) as stream:
        shared_buckets_data: object = yaml.safe_load(stream.read())

    # If the shared bucket name specified in the config.yaml doesn't exist in the dapla team, report an error
    if config_data['shared_bucket'] not in shared_buckets_data['buckets'].keys():
        errorConsole.print(textwrap.dedent(
            f"""


            In the configuration file "{contextual_path}" in the field "shared_bucket" the provided bucket "{config_data['shared_bucket']}" does not exist.

            Existing shared buckets for {environment}:
              {"\n".join(['- ' + bucket for bucket in shared_buckets_data['buckets'].keys()])}


            """
        ))
        sys.exit(1)

    if 'output_columns' in config_data and (diff := set(config_data['pseudo']) - set(config_data['output_columns'])) > 0:
        errorConsole.print(textwrap.dedent(
            f"""


            In the configuration file "{contextual_path}" in the field "output_columns" not all columns
            targeted by pseudo operations are listed in the "output_columns".

            The columns that are missing are:
              {"\n".join(['- ' + column for column in diff])}


            """
        ))
        sys.exit(1)

    # TODO: Check that columns in pseudo tasks aren't targeted more than once

    console.print(f'The "{contextual_path}" configuration was successfully validated!')

if __name__ == '__main__':
  [environment, folder, directory_path, shared_buckets_file_path, *rest] = sys.argv[1:]
  validate_config(environment, folder, directory_path, shared_buckets_file_path)

File: schema-validation/src/validateConfig.scala
//> using scala 3.6.3
//> using dep org.typelevel::cats-core:2.13.0
//> using dep io.circe::circe-yaml:1.15.0
//> using dep io.circe::circe-generic:0.14.13
//> using dep io.circe::circe-parser:0.14.13
//> using dep com.networknt:json-schema-validator:1.5.6
//> using dep ch.qos.logback:logback-classic:1.5.18
//> using dep "dapla-kuben-resource-model:dapla-kuben-resource-model:1.0.3,url=https://github.com/statisticsnorway/dapla-kuben-resource-model/releases/download/java-v1.0.3/dapla-kuben-resource-model-1.0.3.jar"
//> using test.dep org.scalameta::munit::1.1.0
//> using files types.scala configUtils.scala utils.scala
package validate

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.networknt.schema.serialization.JsonNodeReader
import com.networknt.schema.SpecVersion.VersionFlag
import com.networknt.schema.*
import com.schemavalidation.types.{given, *}
import com.schemavalidation.config.utils.*
import com.schemavalidation.utils.*
import io.circe.*
import io.circe.syntax.*
import io.circe.yaml
import java.nio.file.{Files, Path, Paths}
import scala.collection.immutable.*
import scala.io.Source
import scala.jdk.CollectionConverters.*
import scala.util.boundary, boundary.break
import no.ssb.dapla.kuben.v1.{SharedBucket, SharedBuckets}

/** Represents the different possible validation erros that can occur in a
  * delomaten configuration file.
  */
enum ValidationError:
  case SchemaValidationError(
      errorMessages: Set[ValidationMessage]
  )
  case NonExistantSharedBuckets(
      environment: String,
      sharedBucket: String,
      sharedBuckets: List[String]
  )
  case OverlappingPseudoTaskColumns(
      firstTaskName: String,
      secondTaskName: String,
      overlappingColumns: Set[String]
  )

/** Run all validations on the 'config.yaml' file and return a list of all
  * validation errors found.
  *
  * @param configDataPath
  *   the filepath to the 'config.yaml' file containing 'delomaten'
  *   configuration
  * @param contextualPath
  *   an artificial path of the form $environment/$folder/config.yaml used to
  *   provide better error messages
  * @param sharedBucketsPath
  *   the filepath to the shared-buckets buckets-shared.yaml file containing the
  *   teams' shared-buckets
  * @param environment
  *   the dapla team environment the source belongs to i.e. 'uh-varer-prod' or
  *   'play-obr-test'
  * @return
  *   a list of `ValidationError`s if any validation checks fail, otherwise an
  *   empty list
  */
def validateConfiguration(
    configDataPath: Path,
    contextualPath: Path,
    sharedBucketsPath: Path,
    environment: String
): List[ValidationError] =
  import ValidationError.*
  val delomaten: DelomatenConfig =
    loadConfig[DelomatenConfig](configDataPath) match
      case Left(ParsingFailure(msg, _)) =>
        throw Exception(s"Failed to parse delomaten config: $msg".red.newlines)
      case Left(error: DecodingFailure) =>
        throw Exception(
          s"Failed to decode delomaten config: $error".red.newlines
        )
      case Right(config) => config
  val sharedBuckets: SharedBuckets =
    loadConfig[SharedBuckets](sharedBucketsPath) match
      case Left(ParsingFailure(msg, _)) =>
        throw Exception(
          s"Failed to parse shared buckets config: $msg".red.newlines
        )
      case Left(error: DecodingFailure) =>
        throw Exception(
          s"Failed to decode shared buckets config: $error".red.newlines
        )
      case Right(config) => config

  val pseudoTargetedColumns: Set[String] =
    delomaten.pseudo.flatMap(_.columns).toSet

  val schemaValidationErrors = validateConfigSchema(configDataPath) match
    case errMessages if errMessages.nonEmpty =>
      Some(ValidationError.SchemaValidationError(errMessages))
    case _ => None

  // We have to check validation errors first, otherwise the
  // other validations may crash if they can't decode the
  // configuration properly
  schemaValidationErrors match
    case Some(validationError) => List(validationError)
    case None =>
      List(
        if !sharedBuckets
            .getBuckets()
            .asScala
            .toList
            .exists(bucket =>
              delomaten.sharedBucket `contains` bucket.getName()
            )
        then
          Some(
            NonExistantSharedBuckets(
              environment,
              delomaten.sharedBucket,
              sharedBuckets.getBuckets().asScala.toList.map(_.getName())
            )
          )
        else None,
        pseudoTaskColumnsUniquelyTargeted(delomaten.pseudo)
      ).flatten

/** Print error messags for validation errors
  *
  * @param validationErrors
  *   validation errors
  * @return
  *   Unit
  */
def printErrors(
    contextualPath: Path,
    validationErrors: List[ValidationError]
): Unit =
  import ValidationError.*
  validationErrors.foreach { valError =>
    valError match
      case SchemaValidationError(errorMessages) =>
        println(
          s"The delomaten configuration file '$contextualPath' is invalid:\n\n${errorMessages.mkString("\n")}".red.newlines
        )
      case NonExistantSharedBuckets(env, sharedBucket, sharedBuckets) =>
        println(s"""
        |In the configuration file '$contextualPath' in the field 'shared_bucket' the provided bucket '${sharedBucket}' does not exist.

        |Existing shared buckets for $env:
        |  ${sharedBuckets.map("- " + _).mkString("\n  ")}
        |""".stripMargin.red.newlines)
      case OverlappingPseudoTaskColumns(
            firstTaskName,
            secondTaskName,
            overlappingColumns
          ) =>
        println(s"""
        |In the configuration file '$contextualPath' the pseudo tasks '${firstTaskName}' and '${secondTaskName}' target overlapping columns.

        |Overlapping columns:
        |  ${overlappingColumns.map("- " + _).mkString("\n")}
        |""".stripMargin.red.newlines)
  }

/** Validate if the config.yaml under the @directoryPath is valid. Only uses
  * @environment
  *   and @folder arguments to print more informative logs.
  *
  * @param environment
  *   the dapla team environment the source belongs to i.e. 'uh-varer-prod' or
  *   'play-obr-test'
  * @param folder
  *   the statistics product folder which contains the configuration file i.e.
  *   'ledstill' or 'sykefra'
  * @param directoryPath
  *   the filepath to the config.yaml file containing 'delomaten' configuration
  * @param sharedBucketsPathStr
  *   the filepath to the shared-buckets buckets-shared.yaml file containing the
  *   teams' shared-buckets
  * @return
  *   Unit
  */
@main def runValidation(
    environment: String,
    folder: String,
    directoryPath: String,
    sharedBucketsPathStr: String
): Unit =

  if !Files.exists(Paths.get(directoryPath)) then
    println(
      s"The given directory path ${directoryPath} does not exist".red.newlines
    )
    System.exit(1)

  val configDataPath: Path = Paths.get(directoryPath, "config.yaml")
  val contextualPath: Path = Paths.get(environment, folder, "config.yaml")

  if !Files.exists(configDataPath) then
    val context = Paths.get(environment, folder)
    println(
      s"No 'config.yaml' file exists in the product source folder '${context}'".red.newlines
    )
    System.exit(1)

  val sharedBucketsPath: Path = Paths.get(sharedBucketsPathStr)

  if !Files.exists(sharedBucketsPath) then
    println(
      s"The given shared-buckets path '${sharedBucketsPath}' does not exist".red.newlines
    )
    System.exit(1)

  validateConfiguration(
    configDataPath,
    contextualPath,
    sharedBucketsPath,
    environment
  ) match
    case validationErrors if validationErrors.nonEmpty =>
      printErrors(contextualPath, validationErrors)
      System.exit(1)
    case _ => ()

  println(
    s"The '${contextualPath}' configuration was successfully validated!".green.newlines
  )

import ValidationError.OverlappingPseudoTaskColumns
// Ensure that the pseudo task columns are only targeted once.
def pseudoTaskColumnsUniquelyTargeted(
    pseudoTasks: List[PseudoTask]
): Option[OverlappingPseudoTaskColumns] =
  val assocMap: Map[String, List[String]] =
    pseudoTasks.map(task => task.name -> task.columns).toMap

  boundary[Option[OverlappingPseudoTaskColumns]]:
    for (taskName, targetedColumns) <- assocMap do
      val remainingMap = assocMap - taskName
      for (taskNameB, targetedColumnsB) <- remainingMap do
        val overlappingColumns: Set[String] =
          targetedColumns.toSet & targetedColumnsB.toSet
        if overlappingColumns.nonEmpty then
          break[Option[OverlappingPseudoTaskColumns]](
            Some(
              OverlappingPseudoTaskColumns(
                taskName,
                taskNameB,
                overlappingColumns
              )
            )
          )
    None

def validateConfigSchema(filepath: Path): Set[ValidationMessage] =
  val schemaData: String =
    Source.fromFile("./config_schema_spec.yaml").getLines().mkString("\n")
  val inputData: String = Source
    .fromFile(filepath.toFile())
    .getLines()
    .mkString("\n")

  val yamlMapper: ObjectMapper = ObjectMapper(YAMLFactory())
  val jsonNodeReader: JsonNodeReader =
    JsonNodeReader.builder().yamlMapper(yamlMapper).build()
  val config: SchemaValidatorsConfig =
    SchemaValidatorsConfig.builder().errorMessageKeyword("errorMessage").build()
  val schema: JsonSchema = JsonSchemaFactory
    .getInstance(
      VersionFlag.V202012,
      builder => builder.jsonNodeReader(jsonNodeReader).build()
    )
    .getSchema(schemaData, InputFormat.YAML, config)
  schema
    .validate(
      inputData,
      InputFormat.YAML,
      (executionContext) =>
        executionContext.getExecutionConfig().setFormatAssertionsEnabled(true)
    )
    .asScala
    .toSet
